File: python/breadth-first-search/BusRoutes.py
from collections import deque, defaultdict
from typing import List

class Solution:
    def numBusesToDestination(self, routes: List[List[int]], source: int, target: int) -> int:
        stop_to_bus_map = defaultdict(list)
        stop_visited = set()
        bus_visited = set()

        for bus, stops in enumerate(routes):
            for stop in stops:
                stop_to_bus_map[stop].append(bus)

        if source == target:
            return 0
        if source not in stop_to_bus_map:
            return -1

        queue = deque([source])
        stop_visited.add(source)

        buses_taken = 0
        while queue:
            size = len(queue)
            for _ in range(size):
                curr_stop = queue.popleft()
                if curr_stop == target:
                    return buses_taken
                bus_list = stop_to_bus_map[curr_stop]
                for bus in bus_list:
                    if bus not in bus_visited:
                        for stop in routes[bus]:
                            if stop not in stop_visited:
                                queue.append(stop)
                                stop_visited.add(stop)
                        bus_visited.add(bus)
            buses_taken += 1

        return -1

File: java/breadth-first-search/BusRoutes.java
import java.util.*;

class Solution {
    public int numBusesToDestination(int[][] routes, int source, int target) {
        Map<Integer, List<Integer>> stopToBusMap = new HashMap<>();
        Set<Integer> stopVisited = new HashSet<>();
        Set<Integer> busVisited = new HashSet<>();

        for(int bus = 0; bus < routes.length; bus++) {
            for(int stop : routes[bus]) {
                if(!stopToBusMap.containsKey(stop)) {
                    stopToBusMap.put(stop, new ArrayList<>());
                }
                stopToBusMap.get(stop).add(bus);
            }
        }

        if(source == target) return 0;
        if(!stopToBusMap.containsKey(source)) return -1;

        Queue<Integer> queue = new LinkedList<>();
        queue.add(source);
        stopVisited.add(source);

        int busesTaken = 0;
        while(!queue.isEmpty()) {
            int size = queue.size();
            for(int i = 0; i < size; i++) {
                int currStop = queue.poll();
                if(currStop == target) return busesTaken;
                List<Integer> busList = stopToBusMap.get(currStop);
                for(int bus : busList) {
                    if(!busVisited.contains(bus)) {
                        for(int stop : routes[bus]) {
                            if(!stopVisited.contains(stop)) {
                                queue.add(stop);
                                stopVisited.add(stop);
                            }
                        }
                        busVisited.add(bus);
                    }
                }
            }
            busesTaken += 1;
        }
        return -1;
    }
}